src/wallet/gtest/test_wallet.cpp

```diff
@@ -679,6 +697,50 @@ TEST(wallet_tests, ClearNoteWitnessCache) {
     EXPECT_FALSE((bool) witnesses[1]);
 }
 
+TEST(wallet_tests, WriteWitnessCache) {
+    TestWallet wallet;
+    MockWalletDB walletdb;
+
+    auto sk = libzcash::SpendingKey::random();
+    wallet.AddSpendingKey(sk);
+
+    auto wtx = GetValidReceive(sk, 10, true);
+    wallet.AddToWallet(wtx, true, NULL);
+
+    EXPECT_CALL(walletdb, TxnBegin())
+        .WillOnce(Return(false))
+        .WillRepeatedly(Return(true));
+    EXPECT_CALL(walletdb, TxnCommit())
+        .WillOnce(Return(false))
+        .WillRepeatedly(Return(true));
+    EXPECT_CALL(walletdb, TxnAbort())
+        .Times(4);
+
+    EXPECT_CALL(walletdb, WriteTx(wtx.GetHash(), wtx))
+        .WillOnce(Return(false))
+        .WillOnce(ThrowLogicError())
+        .WillRepeatedly(Return(true));
+    EXPECT_CALL(walletdb, WriteWitnessCacheSize(0))
+        .WillOnce(Return(false))
+        .WillOnce(ThrowLogicError())
+        .WillRepeatedly(Return(true));
```

Comment:
I find this style of test very difficult to follow; split it up so that the EXPECT_CALLs are associated with a given call to ``WriteWitnessCache``. Otherwise it's not possible to tell, for instance, *which* calls to ``WriteWitnessCache`` called ``TxnAbort``.